File: frontend/src/api/commandCenter.js
/**
 * Command Center API client
 * Provides access to market data, prediction markets, news, and AI copilot
 */
import apiClient from './axios';

const BASE_URL = '/api/v1/command-center';

// =============================================================================
// DASHBOARD
// =============================================================================

/**
 * Get full dashboard data in a single request
 * Optimized for initial page load
 */
export const getDashboard = async () => {
  const response = await apiClient.get(`${BASE_URL}/dashboard`);
  return response.data;
};

// =============================================================================
// MARKET DATA
// =============================================================================

/**
 * Get complete market summary
 * Includes indices, volatility, Fear & Greed, and sectors
 */
export const getMarketSummary = async () => {
  const response = await apiClient.get(`${BASE_URL}/market/summary`);
  return response.data;
};

/**
 * Get market indices (SPY, QQQ, DIA, IWM)
 */
export const getMarketIndices = async () => {
  const response = await apiClient.get(`${BASE_URL}/market/indices`);
  return response.data;
};

/**
 * Get volatility metrics (VIX, VIX9D)
 */
export const getVolatility = async () => {
  const response = await apiClient.get(`${BASE_URL}/market/volatility`);
  return response.data;
};

/**
 * Get CNN Fear & Greed Index
 */
export const getFearGreed = async () => {
  const response = await apiClient.get(`${BASE_URL}/market/fear-greed`);
  return response.data;
};

/**
 * Get sector ETF performance
 */
export const getSectorPerformance = async () => {
  const response = await apiClient.get(`${BASE_URL}/market/sectors`);
  return response.data;
};

// =============================================================================
// POLYMARKET
// =============================================================================

/**
 * Get Polymarket dashboard summary
 */
export const getPolymarketDashboard = async () => {
  const response = await apiClient.get(`${BASE_URL}/polymarket/dashboard`);
  return response.data;
};

/**
 * Get trading-relevant prediction markets
 * @param {number} limit - Number of markets to return
 */
export const getTradingMarkets = async (limit = 20) => {
  const response = await apiClient.get(`${BASE_URL}/polymarket/markets`, {
    params: { limit },
  });
  return response.data;
};

/**
 * Get markets with significant odds changes
 * @param {number} threshold - Minimum change percentage
 */
export const getSignificantChanges = async (threshold = 5.0) => {
  const response = await apiClient.get(`${BASE_URL}/polymarket/changes`, {
    params: { threshold },
  });
  return response.data;
};

/**
 * Get curated key markets by category
 */
export const getKeyMarkets = async () => {
  const response = await apiClient.get(`${BASE_URL}/polymarket/key-markets`);
  return response.data;
};

// =============================================================================
// NEWS FEED
// =============================================================================

/**
 * Get general market news
 * @param {number} limit - Number of news items
 */
export const getMarketNews = async (limit = 20) => {
  const response = await apiClient.get(`${BASE_URL}/news/market`, {
    params: { limit },
  });
  return response.data;
};

/**
 * Get news for a specific symbol
 * @param {string} symbol - Stock ticker
 * @param {number} limit - Number of news items
 */
export const getCompanyNews = async (symbol, limit = 10) => {
  const response = await apiClient.get(`${BASE_URL}/news/company/${symbol}`, {
    params: { limit },
  });
  return response.data;
};

/**
 * Get economic calendar
 * @param {number} daysAhead - Days to look ahead
 */
export const getEconomicCalendar = async (daysAhead = 7) => {
  const response = await apiClient.get(`${BASE_URL}/news/economic-calendar`, {
    params: { days_ahead: daysAhead },
  });
  return response.data;
};

/**
 * Get earnings calendar
 * @param {number} daysAhead - Days to look ahead
 */
export const getEarningsCalendar = async (daysAhead = 14) => {
  const response = await apiClient.get(`${BASE_URL}/news/earnings-calendar`, {
    params: { days_ahead: daysAhead },
  });
  return response.data;
};

/**
 * Get combined catalyst feed
 * @param {number} limit - Number of items
 */
export const getCatalystFeed = async (limit = 20) => {
  const response = await apiClient.get(`${BASE_URL}/news/catalyst-feed`, {
    params: { limit },
  });
  return response.data;
};

// =============================================================================
// AI COPILOT
// =============================================================================

/**
 * Get AI-generated morning brief
 */
export const getMorningBrief = async () => {
  const response = await apiClient.get(`${BASE_URL}/copilot/morning-brief`);
  return response.data;
};

/**
 * Get AI explanation for a metric
 * @param {string} metricName - Name of the metric
 * @param {any} metricValue - Current value
 * @param {string} context - Additional context
 */
export const explainMetric = async (metricName, metricValue, context = '') => {
  const response = await apiClient.post(`${BASE_URL}/copilot/explain`, {
    metric_name: metricName,
    metric_value: metricValue,
    context,
  });
  return response.data;
};

/**
 * Chat with the AI copilot
 * @param {string} message - User message
 * @param {Object} context - Current context
 * @param {Array} conversationHistory - Previous messages
 */
export const chatWithCopilot = async (message, context = null, conversationHistory = null) => {
  const response = await apiClient.post(`${BASE_URL}/copilot/chat`, {
    message,
    context,
    conversation_history: conversationHistory,
  });
  return response.data;
};

/**
 * Get AI analysis for a stock
 * @param {Object} stockData - Stock data from detail page
 * @param {Object} marketContext - Current market context
 */
export const analyzeStock = async (stockData, marketContext = null) => {
  const response = await apiClient.post(`${BASE_URL}/copilot/analyze-stock`, {
    stock_data: stockData,
    market_context: marketContext,
  });
  return response.data;
};

/**
 * Get AI copilot status and usage
 */
export const getCopilotStatus = async () => {
  const response = await apiClient.get(`${BASE_URL}/copilot/status`);
  return response.data;
};

// =============================================================================
// MACRO SIGNALS
// =============================================================================

/**
 * Get current Macro Risk Index (MRI)
 */
export const getMRI = async () => {
  const response = await apiClient.get(`${BASE_URL}/macro/mri`);
  return response.data;
};

/**
 * Get MRI history
 * @param {number} hours - Hours of history to retrieve
 */
export const getMRIHistory = async (hours = 24) => {
  const response = await apiClient.get(`${BASE_URL}/macro/mri/history`, {
    params: { hours },
  });
  return response.data;
};

/**
 * Get MRI components breakdown
 */
export const getMRIComponents = async () => {
  const response = await apiClient.get(`${BASE_URL}/macro/mri/components`);
  return response.data;
};

/**
 * Get macro events/prediction markets
 * @param {number} limit - Number of markets to return
 */
export const getMacroEvents = async (limit = 50) => {
  const response = await apiClient.get(`${BASE_URL}/macro/events`, {
    params: { limit },
  });
  return response.data;
};

/**
 * Get macro events by category
 * @param {string} category - Category filter
 */
export const getMacroEventsByCategory = async (category) => {
  const response = await apiClient.get(`${BASE_URL}/macro/events/category/${category}`);
  return response.data;
};

/**
 * Get macro divergences
 */
export const getMacroDivergences = async () => {
  const response = await apiClient.get(`${BASE_URL}/macro/divergences`);
  return response.data;
};

/**
 * Get narrative momentum
 * @param {string} category - Optional category filter
 * @param {number} threshold - Minimum change threshold
 */
export const getMacroMomentum = async (category = null, threshold = 10.0) => {
  const params = { threshold };
  if (category) params.category = category;
  const response = await apiClient.get(`${BASE_URL}/macro/momentum`, { params });
  return response.data;
};

/**
 * Get ticker macro bias
 * @param {string} symbol - Stock ticker
 */
export const getTickerMacroBias = async (symbol) => {
  const response = await apiClient.get(`${BASE_URL}/macro/ticker/${symbol}/bias`);
  return response.data;
};

/**
 * Get macro signal health status
 */
export const getMacroHealth = async () => {
  const response = await apiClient.get(`${BASE_URL}/macro/health`);
  return response.data;
};

export default {
  // Dashboard
  getDashboard,
  // Market Data
  getMarketSummary,
  getMarketIndices,
  getVolatility,
  getFearGreed,
  getSectorPerformance,
  // Polymarket
  getPolymarketDashboard,
  getTradingMarkets,
  getSignificantChanges,
  getKeyMarkets,
  // News
  getMarketNews,
  getCompanyNews,
  getEconomicCalendar,
  getEarningsCalendar,
  getCatalystFeed,
  // Copilot
  getMorningBrief,
  explainMetric,
  chatWithCopilot,
  analyzeStock,
  getCopilotStatus,
  // Macro Signals
  getMRI,
  getMRIHistory,
  getMRIComponents,
  getMacroEvents,
  getMacroEventsByCategory,
  getMacroDivergences,
  getMacroMomentum,
  getTickerMacroBias,
  getMacroHealth,
};

File: backend/app/api/endpoints/command_center.py
"""
Command Center API Endpoints
Provides market data, news, prediction markets, and AI copilot functionality
"""

from fastapi import APIRouter, HTTPException, Query
from pydantic import BaseModel, Field
from typing import Optional, List, Dict, Any
from datetime import datetime
from loguru import logger

from app.services.command_center import (
    get_market_data_service,
    get_polymarket_service,
    get_news_feed_service,
    get_news_service,
    get_copilot_service,
)

router = APIRouter()


# =============================================================================
# REQUEST/RESPONSE MODELS
# =============================================================================

class ChatMessage(BaseModel):
    """Chat message model"""
    role: str = Field(..., description="Message role: 'user' or 'assistant'")
    content: str = Field(..., description="Message content")


class ChatRequest(BaseModel):
    """Request model for copilot chat"""
    message: str = Field(..., description="User message")
    context: Optional[Dict[str, Any]] = Field(None, description="Current context (selected stock, page, etc.)")
    conversation_history: Optional[List[ChatMessage]] = Field(None, description="Previous messages")


class ExplainMetricRequest(BaseModel):
    """Request model for metric explanation"""
    metric_name: str = Field(..., description="Name of the metric to explain")
    metric_value: Any = Field(..., description="Current value of the metric")
    context: Optional[str] = Field(None, description="Additional context")


class StockAnalysisRequest(BaseModel):
    """Request model for stock detail analysis"""
    stock_data: Dict[str, Any] = Field(..., description="Stock data from the detail page")
    market_context: Optional[Dict[str, Any]] = Field(None, description="Current market context")


# =============================================================================
# MARKET DATA ENDPOINTS
# =============================================================================

@router.get("/market/summary")
async def get_market_summary():
    """
    Get complete market summary for the Command Center.
    Includes indices, volatility, Fear & Greed, and sector performance.
    """
    try:
        service = get_market_data_service()
        summary = await service.get_market_summary()
        return summary
    except Exception as e:
        logger.error(f"Error getting market summary: {e}")
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/market/indices")
async def get_market_indices():
    """
    Get current market indices (SPY, QQQ, DIA, IWM).
    """
    try:
        service = get_market_data_service()
        indices = await service.get_market_indices()
        return {"indices": indices, "timestamp": datetime.now().isoformat()}
    except Exception as e:
        logger.error(f"Error getting market indices: {e}")
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/market/volatility")
async def get_volatility_metrics():
    """
    Get VIX and volatility metrics.
    """
    try:
        service = get_market_data_service()
        volatility = await service.get_volatility_metrics()
        return {"volatility": volatility, "timestamp": datetime.now().isoformat()}
    except Exception as e:
        logger.error(f"Error getting volatility metrics: {e}")
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/market/fear-greed")
async def get_fear_greed():
    """
    Get CNN Fear & Greed Index.
    """
    try:
        service = get_market_data_service()
        fear_greed = await service.get_fear_greed_index()
        if fear_greed:
            return fear_greed
        else:
            raise HTTPException(status_code=503, detail="Fear & Greed data unavailable")
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Error getting Fear & Greed: {e}")
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/market/sectors")
async def get_sector_performance():
    """
    Get sector ETF performance for rotation analysis.
    """
    try:
        service = get_market_data_service()
        sectors = await service.get_sector_performance()
        return {"sectors": sectors, "timestamp": datetime.now().isoformat()}
    except Exception as e:
        logger.error(f"Error getting sector performance: {e}")
        raise HTTPException(status_code=500, detail=str(e))


# =============================================================================
# POLYMARKET ENDPOINTS
# =============================================================================

@router.get("/polymarket/dashboard")
async def get_polymarket_dashboard():
    """
    Get Polymarket data formatted for the dashboard widget.
    Includes key markets and significant changes.
    """
    try:
        service = get_polymarket_service()
        dashboard = await service.get_dashboard_summary()
        return dashboard
    except Exception as e:
        logger.error(f"Error getting Polymarket dashboard: {e}")
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/polymarket/markets")
async def get_trading_markets(
    limit: int = Query(20, ge=1, le=50, description="Number of markets to return"),
):
    """
    Get all trading-relevant prediction markets.
    """
    try:
        service = get_polymarket_service()
        markets = await service.get_trading_markets(limit=limit)
        return {"markets": markets, "timestamp": datetime.now().isoformat()}
    except Exception as e:
        logger.error(f"Error getting Polymarket markets: {e}")
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/polymarket/changes")
async def get_significant_changes(
    threshold: float = Query(5.0, ge=1.0, le=20.0, description="Minimum change percentage"),
):
    """
    Get markets with significant odds changes for alerts.
    """
    try:
        service = get_polymarket_service()
        changes = await service.get_significant_changes(threshold=threshold)
        return {"changes": changes, "timestamp": datetime.now().isoformat()}
    except Exception as e:
        logger.error(f"Error getting Polymarket changes: {e}")
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/polymarket/key-markets")
async def get_key_markets():
    """
    Get curated key markets by category.
    """
    try:
        service = get_polymarket_service()
        key_markets = await service.get_key_markets()
        return key_markets
    except Exception as e:
        logger.error(f"Error getting key markets: {e}")
        raise HTTPException(status_code=500, detail=str(e))


# =============================================================================
# NEWS FEED ENDPOINTS
# =============================================================================

@router.get("/news/market")
async def get_market_news(
    limit: int = Query(20, ge=1, le=50, description="Number of news items"),
):
    """
    Get general market news aggregated from multiple sources.
    """
    try:
        service = get_news_feed_service()
        news = await service.get_market_news(limit=limit)
        return {"news": news, "timestamp": datetime.now().isoformat()}
    except Exception as e:
        logger.error(f"Error getting market news: {e}")
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/news/company/{symbol}")
async def get_company_news(
    symbol: str,
    limit: int = Query(10, ge=1, le=30, description="Number of news items"),
):
    """
    Get news for a specific company/symbol.
    """
    try:
        service = get_news_feed_service()
        news = await service.get_company_news(symbol=symbol.upper(), limit=limit)
        return {"symbol": symbol.upper(), "news": news, "timestamp": datetime.now().isoformat()}
    except Exception as e:
        logger.error(f"Error getting company news for {symbol}: {e}")
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/news/economic-calendar")
async def get_economic_calendar(
    days_ahead: int = Query(7, ge=1, le=30, description="Days to look ahead"),
):
    """
    Get upcoming economic events.
    """
    try:
        service = get_news_feed_service()
        events = await service.get_economic_calendar(days_ahead=days_ahead)
        return {"events": events, "timestamp": datetime.now().isoformat()}
    except Exception as e:
        logger.error(f"Error getting economic calendar: {e}")
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/news/earnings-calendar")
async def get_earnings_calendar(
    days_ahead: int = Query(14, ge=1, le=30, description="Days to look ahead"),
):
    """
    Get upcoming earnings announcements.
    """
    try:
        service = get_news_feed_service()
        earnings = await service.get_earnings_calendar(days_ahead=days_ahead)
        return {"earnings": earnings, "timestamp": datetime.now().isoformat()}
    except Exception as e:
        logger.error(f"Error getting earnings calendar: {e}")
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/news/catalyst-feed")
async def get_catalyst_feed(
    limit: int = Query(20, ge=1, le=50, description="Number of items"),
):
    """
    Get combined catalyst feed for the Command Center.
    Combines news, economic events, and earnings into prioritized feed.
    """
    try:
        service = get_news_feed_service()
        feed = await service.get_catalyst_feed(limit=limit)
        return feed
    except Exception as e:
        logger.error(f"Error getting catalyst feed: {e}")
        raise HTTPException(status_code=500, detail=str(e))


# =============================================================================
# RSS NEWS FEED ENDPOINTS (Free - No API Key Required)
# =============================================================================

@router.get("/news/rss")
async def get_rss_news(
    limit: int = Query(20, ge=1, le=50, description="Number of news items"),
    category: Optional[str] = Query(None, description="Filter by category: markets, investing, stocks, general"),
):
    """
    Get aggregated financial news from RSS feeds (CNBC, Yahoo Finance, MarketWatch, etc.).
    No API key required.
    """
    try:
        service = get_news_service()
        news = await service.get_news(limit=limit, category=category)
        return {"news": news, "timestamp": datetime.now().isoformat(), "source": "rss_feeds"}
    except Exception as e:
        logger.error(f"Error getting RSS news: {e}")
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/news/rss/market")
async def get_rss_market_news(
    limit: int = Query(15, ge=1, le=50, description="Number of news items"),
):
    """
    Get market-focused news filtered by financial keywords.
    Sources: CNBC, Yahoo Finance, MarketWatch, Reuters, Investing.com
    """
    try:
        service = get_news_service()
        news = await service.get_market_news(limit=limit)
        return {"news": news, "timestamp": datetime.now().isoformat(), "source": "rss_feeds"}
    except Exception as e:
        logger.error(f"Error getting RSS market news: {e}")
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/news/rss/sources")
async def get_rss_sources():
    """
    Get list of available RSS news sources.
    """
    try:
        service = get_news_service()
        sources = service.get_available_sources()
        return {"sources": sources, "timestamp": datetime.now().isoformat()}
    except Exception as e:
        logger.error(f"Error getting RSS sources: {e}")
        raise HTTPException(status_code=500, detail=str(e))


# =============================================================================
# AI COPILOT ENDPOINTS
# =============================================================================

@router.get("/copilot/morning-brief")
async def get_morning_brief():
    """
    Get AI-generated morning brief based on current market conditions.
    """
    try:
        copilot = get_copilot_service()
        market_service = get_market_data_service()
        polymarket_service = get_polymarket_service()
        news_service = get_news_feed_service()

        # Gather data for brief
        market_data = await market_service.get_market_summary()
        polymarket_data = await polymarket_service.get_dashboard_summary()
        news_data = await news_service.get_catalyst_feed(limit=10)

        # Generate brief
        brief = await copilot.generate_morning_brief(
            market_data=market_data,
            polymarket_data=polymarket_data,
            news_data=news_data,
        )

        return brief

    except Exception as e:
        logger.error(f"Error generating morning brief: {e}")
        raise HTTPException(status_code=500, detail=str(e))


@router.post("/copilot/explain")
async def explain_metric(request: ExplainMetricRequest):
    """
    Get AI explanation for a specific metric.
    Used for hover tooltips and learning features.
    """
    try:
        copilot = get_copilot_service()
        explanation = await copilot.explain_metric(
            metric_name=request.metric_name,
            metric_value=request.metric_value,
            context=request.context or "",
        )
        return explanation

    except Exception as e:
        logger.error(f"Error explaining metric: {e}")
        raise HTTPException(status_code=500, detail=str(e))


@router.post("/copilot/chat")
async def copilot_chat(request: ChatRequest):
    """
    Interactive chat with the AI copilot.
    Supports context-aware conversations about market data.
    """
    try:
        copilot = get_copilot_service()

        # Convert Pydantic models to dicts for conversation history
        history = None
        if request.conversation_history:
            history = [{"role": m.role, "content": m.content} for m in request.conversation_history]

        response = await copilot.chat(
            message=request.message,
            context=request.context,
            conversation_history=history,
        )

        return response

    except Exception as e:
        logger.error(f"Error in copilot chat: {e}")
        raise HTTPException(status_code=500, detail=str(e))


@router.post("/copilot/analyze-stock")
async def analyze_stock_detail(request: StockAnalysisRequest):
    """
    Generate comprehensive AI analysis for a stock detail page.
    """
    try:
        copilot = get_copilot_service()
        analysis = await copilot.analyze_stock_detail(
            stock_data=request.stock_data,
            market_context=request.market_context,
        )
        return analysis

    except Exception as e:
        logger.error(f"Error analyzing stock: {e}")
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/copilot/status")
async def get_copilot_status():
    """
    Get AI copilot status and usage statistics.
    """
    try:
        copilot = get_copilot_service()

        return {
            "available": copilot.is_available(),
            "usage": copilot.get_usage_stats(),
            "timestamp": datetime.now().isoformat(),
        }

    except Exception as e:
        logger.error(f"Error getting copilot status: {e}")
        raise HTTPException(status_code=500, detail=str(e))


# =============================================================================
# COMBINED DASHBOARD ENDPOINT
# =============================================================================

@router.get("/dashboard")
async def get_full_dashboard():
    """
    Get all Command Center data in a single request.
    Optimized for initial page load.
    """
    try:
        market_service = get_market_data_service()
        polymarket_service = get_polymarket_service()
        news_service = get_news_feed_service()
        rss_news_service = get_news_service()
        copilot = get_copilot_service()

        # Fetch all data concurrently
        import asyncio

        market_task = asyncio.create_task(market_service.get_market_summary())
        polymarket_task = asyncio.create_task(polymarket_service.get_dashboard_summary())
        catalyst_task = asyncio.create_task(news_service.get_catalyst_feed(limit=15))
        rss_news_task = asyncio.create_task(rss_news_service.get_market_news(limit=15))

        market_data = await market_task
        polymarket_data = await polymarket_task
        catalyst_data = await catalyst_task
        rss_news_data = await rss_news_task

        # Generate morning brief (if AI available)
        brief = None
        if copilot.is_available():
            try:
                brief_result = await copilot.generate_morning_brief(
                    market_data=market_data,
                    polymarket_data=polymarket_data,
                    news_data=catalyst_data,
                )
                if brief_result.get('success'):
                    brief = brief_result.get('brief')
            except Exception as e:
                logger.warning(f"Could not generate morning brief: {e}")

        return {
            "timestamp": datetime.now().isoformat(),
            "market": market_data,
            "polymarket": polymarket_data,
            "catalysts": catalyst_data,
            "rss_news": rss_news_data,
            "morning_brief": brief,
            "copilot_available": copilot.is_available(),
        }

    except Exception as e:
        logger.error(f"Error getting full dashboard: {e}")
        raise HTTPException(status_code=500, detail=str(e))
